dbms/src/Flash/Coprocessor/FunctionBuildHelper.cpp

```diff
@@ -72,62 +72,69 @@ const std::unordered_map<String, String> DateSub::unit_to_func_name_map
-        String name = analyzer->getActions(expr.children(i), actions, output_as_uint8_type);
+        bool output_as_uint8_type = (i + 1) != function_expr.argumentSize() && (i % 2 == 0);
+        String name = function_expr.buildArgument(analyzer, actions, i, output_as_uint8_type);
         argument_names.push_back(name);
     }
     if (argument_names.size() % 2 == 0)
     {
         String name = analyzer->getActions(constructNULLLiteralTiExpr(), actions);
         argument_names.push_back(name);
     }
-    return analyzer->applyFunction(func_name, argument_names, actions, getCollatorFromExpr(expr));
+    return analyzer->applyFunction(func_name, argument_names, actions, function_expr.getFunctionCollator());
 }
 
-String DAGExpressionAnalyzerHelper::buildIfNullFunction(
+String FunctionBuildHelper::buildIfNullFunction(
+    const FunctionExpr & function_expr,
     DAGExpressionAnalyzer * analyzer,
-    const tipb::Expr & expr,
     const ExpressionActionsPtr & actions)
 {
     // rewrite IFNULL function with multiIf
     // ifNull(arg1, arg2) -> multiIf(isNull(arg1), arg2, assumeNotNull(arg1))
     // todo if arg1 is not nullable, then just return arg1 is ok
     const String & func_name = "multiIf";
     Names argument_names;
-    if (expr.children_size() != 2)
+    if (function_expr.argumentSize() != 2)
     {
         throw TiFlashException("Invalid arguments of IFNULL function", Errors::Coprocessor::BadRequest);
     }
 
-    String condition_arg_name = analyzer->getActions(expr.children(0), actions, false);
-    String else_arg_name = analyzer->getActions(expr.children(1), actions, false);
-    String is_null_result = analyzer->applyFunction("isNull", {condition_arg_name}, actions, getCollatorFromExpr(expr));
+    String condition_arg_name = function_expr.buildArgument(analyzer, actions, 0, false);
+    String else_arg_name = function_expr.buildArgument(analyzer, actions, 1, false);
+    String is_null_result = analyzer->applyFunction("isNull", {condition_arg_name}, actions, function_expr.getFunctionCollator());
     String not_null_condition_arg_name = analyzer->applyFunction("assumeNotNull", {condition_arg_name}, actions, nullptr);
 
     argument_names.push_back(std::move(is_null_result));
     argument_names.push_back(std::move(else_arg_name));
     argument_names.push_back(std::move(not_null_condition_arg_name));
 
-    return analyzer->applyFunction(func_name, argument_names, actions, getCollatorFromExpr(expr));
+    return analyzer->applyFunction(func_name, argument_names, actions, function_expr.getFunctionCollator());
 }
 
-String DAGExpressionAnalyzerHelper::buildInFunction(
+String FunctionBuildHelper::buildInFunction(
+    const FunctionExpr & function_expr,
     DAGExpressionAnalyzer * analyzer,
-    const tipb::Expr & expr,
     const ExpressionActionsPtr & actions)
 {
+    const auto * tipb_function_expr = dynamic_cast<const TiPBExprFunctionExpr *>(&function_expr);
```

Comment:
nit: Maybe use NAME to identify the class is better?

Reply:
You mean use NAME instead of `auto`?

Reply:
> You mean use NAME instead of `auto`?
Use name intead of the dynamic cast.